services/venice-controller/src/main/java/com/linkedin/venice/pushmonitor/AbstractPushMonitor.java

```diff
@@ -910,8 +927,136 @@ public void onRoutingDataDeleted(String kafkaTopic) {
+              }
+              if (newStatusDetails.length() > 0) {
+                newStatusDetails.append(", ");
+              }
+              newStatusDetails.append("broadcast EOP to materialized view topic: ").append(materializedViewTopicName);
+            }
+          }
+        }
+        updatePushStatus(
+            offlinePushStatus,
+            ExecutionStatus.END_OF_PUSH_RECEIVED,
+            Optional.of(newStatusDetails.toString()));
+        LOGGER.info("Successfully {} for offlinePushStatus: {}", newStatusDetails.toString(), offlinePushStatus);
+      }
+    } catch (Exception e) {
+      String newStatusDetails = "Failed to start EOP procedures";
+      handleTerminalOfflinePushUpdate(offlinePushStatus, new ExecutionStatusWithDetails(ERROR, newStatusDetails));
+      LOGGER.error("{} for offlinePushStatus: {}", newStatusDetails, offlinePushStatus, e);
+    }
+  }
+
+  /**
+   * For hybrid stores we use {@link com.linkedin.venice.writer.LeaderCompleteState} to prevent follower replicas from
+   * reporting COMPLETED prematurely. The same safeguard is needed for view topic consumers. Due to the obfuscated
+   * partition mapping between version topic leaders and view topic partitions we can only send LEADER_COMPLETED once
+   * all version topic partition leaders are completed. This is equivalent to at least one completed replica for all
+   * partitions.
+   */
+  protected void checkWhetherToStartLeaderCompletedProcedures(OfflinePushStatus offlinePushStatus) {
+    if (topicToLeaderCompleteTimestampMap.containsKey(offlinePushStatus.getKafkaTopic())) {
+      // We've sent the LeaderCompleteState for all view topics of this version topic already. Duplicate heartbeats are
+      // harmless, so we are only tracking it in memory (not across restarts) to avoid heartbeat spam
+      return;
+    }
+    Store store = getStoreOrThrow(offlinePushStatus);
+    Version version = getVersionOrThrow(store, offlinePushStatus.getKafkaTopic());
+    Map<String, ViewConfig> viewConfigMap = version.getViewConfigs();
+    if (!version.isHybrid() || viewConfigMap == null || viewConfigMap.isEmpty()) {
+      return;
+    }
+    if (offlinePushStatus.isLeaderCompletedInEveryPartition()) {
+      // broadcast heartbeat with LeaderCompleteState to view topic partitions
+      long heartbeatTimestamp = System.currentTimeMillis();
+      for (ViewConfig rawViewConfig: viewConfigMap.values()) {
+        if (MaterializedView.class.getCanonicalName().equals(rawViewConfig.getViewClassName())) {
+          VeniceView veniceView = ViewUtils.getVeniceView(
+              rawViewConfig.getViewClassName(),
+              new Properties(),
+              store.getName(),
+              rawViewConfig.getViewParameters());
+          MaterializedView materializedView = (MaterializedView) veniceView;
+          for (String materializedViewTopicName: materializedView.getTopicNamesAndConfigsForVersion(version.getNumber())
+              .keySet()) {
```

Comment:
I see a few places to construct all the view topics belonging to a given store version, so can we build a utility to get rid of the duplicate codes?